File: P3/civitasruby/lib/civitas/dado.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
require 'singleton'

module Civitas
  class Dado
    include Singleton
    
    attr_reader :ultimo_resultado
    
    @@salida_carcel = 5
    
    def initialize
      @random = Random.new
      @ultimo_resultado 
      @debug = false
    end
    
    def tirar
      tirada = @random.rand(6)+1
      if @debug
        tirada = 1
      end
      @ultimo_resultado = tirada
      return tirada
    end
    
    def salgo_de_la_carcel
      tirada = tirar
      @ultimo_resultado = tirada
      if tirada >= @@salida_carcel
        return true
      else
        return false
      end
    end
    
    def quien_empieza(n)
      return @random.rand(n)
    end
    
    def set_debug(d)
      @debug = d
      if @debug
        Diario.instance.ocurre_evento("Activado modo debug de Dado")
      else
        Diario.instance.ocurre_evento("Desactivado modo debug del Dado")
      end
    end
    
    private :initialize
  end
end

File: ExamenParcial1/Civitas/src/civitas/Dado.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;

import java.util.Random;

/**
 *
 * @author salva
 */
public class Dado {

    private Random random;
    private int ultimoResultado;
    private boolean debug;

    static final private Dado instance = new Dado();
    static final private int SalidaCarcel = 5;

    private Dado() {
        debug = false;
        random = new Random();
    }
// EXAMEN
    //hacemos publica la visibilidad para usarla en el paquete exammen

    public static Dado getInstance() {
        return instance;
    }
// FIN DE EXAMEN

    int tirar() {
        int tirada = 0;
        if (debug) {
            tirada = 1;
        } else {
            tirada = random.nextInt(6) + 1;
        }
        ultimoResultado = tirada;
        return tirada;
    }

    boolean salgoDeLaCarcel() {
        int tirada = tirar();
        if (tirada >= SalidaCarcel) {
            ultimoResultado = tirada;
        }
        return tirada >= SalidaCarcel;
    }

    int quienEmpieza(int n) {
        return random.nextInt(n);
    }
// EXAMEN
    //hacemos publica la visibilidad para usarla en el paquete exammen

    public void setDebug(Boolean d) {
        debug = d;
        if (debug) {
            Diario.getInstance().ocurreEvento("Activado modo debug del dado");
        } else {
            Diario.getInstance().ocurreEvento("Desactivado modo debug del dado");
        }
    }
// FIN DE EXAMEN

    int getUltimoResultado() {
        return ultimoResultado;
    }
    // FIN DE EXAMEN
}
